presto-raptor/src/main/java/com/facebook/presto/raptor/storage/InplaceShardRewriter.java

```diff
@@ -0,0 +1,136 @@
+import com.facebook.presto.raptor.metadata.ShardInfo;
+import com.facebook.presto.raptor.metadata.ShardRecorder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.airlift.concurrent.MoreFutures.getFutureValue;
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+
+public class InplaceShardRewriter
+        implements ShardRewriter
+{
+    private static final JsonCodec<ShardDelta> SHARD_DELTA_CODEC = jsonCodec(ShardDelta.class);
+    private final OrcStorageManager orcStorageManager;
+    private final StorageService storageService;
+    private final ShardRecorder shardRecorder;
+    private final BackupManager backupManager;
+    private final FileSystem fileSystem;
+    private final long transactionId;
+    private final OptionalInt bucketNumber;
+    private final UUID shardUuid;
+    private final Map<String, Type> columns;
+    private final ExecutorService deletionExecutor;
+    private final String nodeId;
+
+    public InplaceShardRewriter(
+            FileSystem fileSystem,
+            long transactionId,
+            OptionalInt bucketNumber,
+            UUID shardUuid,
+            Map<String, Type> columns,
+            ExecutorService deletionExecutor,
+            String nodeId,
+            OrcStorageManager orcStorageManager,
+            StorageService storageService,
+            ShardRecorder shardRecorder,
+            BackupManager backupManager)
+    {
+        this.orcStorageManager = orcStorageManager;
+        this.storageService = storageService;
+        this.shardRecorder = shardRecorder;
+        this.backupManager = backupManager;
+        this.fileSystem = fileSystem;
+        this.transactionId = transactionId;
+        this.bucketNumber = bucketNumber;
+        this.shardUuid = shardUuid;
+        this.columns = columns;
+        this.deletionExecutor = deletionExecutor;
+        this.nodeId = nodeId;
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> rewrite(BitSet rowsToDelete)
+    {
+        if (rowsToDelete.isEmpty()) {
+            return completedFuture(ImmutableList.of());
+        }
+        return supplyAsync(() -> rewriteShard(rowsToDelete), deletionExecutor);
+    }
+
+    @VisibleForTesting
```

Comment:
Why don't we just reuse `CompletableFuture<Collection<Slice>> rewrite(BitSet rowsToDelete)` for testing? There is no much difference.